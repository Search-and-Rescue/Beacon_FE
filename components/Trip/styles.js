const styles = {
  tripContainer: {
    flex: 1,
    alignItems: "center"
  },
  tripsList: {
    borderWidth: 2,
    borderColor: "black",
    flex: 1,
    height: 300,
    width: "90%"
  },
  pickerView: {
    backgroundColor: '#efefef',
    margin: 5,
    padding: 20,
    position: 'absolute',
    bottom: 30,
    left: 10,
    right: 10, 
    height: 400
  },
  modalHeading: {
    fontWeight: 'bold',
    marginBottom: 5
  },
  modalButton: {
    paddingVertical: 10,
    marginVertical: 3
  },
  modalToggleGearContainer: {
    flex: 1, 
    flexDirection: 'row'
  },
  modalToggleGearBtn: {
    borderWidth: 1,
    borderColor: 'black',
    marginRight: 10,
    marginVertical: 5,
    paddingVertical: 3,
    textAlign: 'center',
    width: 75
  },
  modalToggleGearHeading: {
    paddingVertical: 10
  },
  modalToggleContactsContainer: {
    flex: 1,
    flexDirection: 'row'
  },
  modalToggleContactsBtn: {
    borderWidth: 1,
    borderColor: 'black',
    marginRight: 10,
    marginVertical: 5,
    paddingVertical: 3,
    textAlign: 'center',
    width: 75
  },
  modalToggleContactsHeading: {
    paddingVertical: 10
  },
  input: {
    backgroundColor: "#fff",
    borderColor: "black",
    borderWidth: 1,
    marginBottom: 10,
    padding: 10,
    width: "80%"
  },
  updateBtn: {
    borderColor: "black",
    borderWidth: 1,
    marginVertical: 3,
    padding: 10,
    textAlign: 'center',
    width: "100%"
  },
  updateBtnText: {
    textAlign: "center"
  }
};

export default styles;
