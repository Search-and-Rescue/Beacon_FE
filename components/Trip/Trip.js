import React, { Component } from "react";
import {
  Button,
  Modal,
  Text,
  TextInput,
  TouchableHighlight,
  TouchableOpacity,
  View
} from "react-native";
import { connect } from "react-redux";
import styles from "./styles";
import { ScrollView } from "react-native-gesture-handler";
import { addTrip, getTrips, addGearForTrip, addContactsForTrip, addVehiclesForTrip } from '../../util/apiCalls';
import { setTrips, setCurrentTrip } from '../../actions';
import { bindActionCreators } from 'redux';

export class Trip extends Component {
  constructor(props) {
    super(props);
    this.state = {
      contacts: [],
      contacts_modal: false,
      vehicle: 0,
      vehicle_modal: false,
      gear: [],
      gear_modal: false,
      name: "",
      startingPoint: "",
      endingPoint: "",
      startDate: "",
      startTime: "",
      endDate: "",
      endTime: "",
      notificationDate: "",
      notificationTime: "",
      travelingCompanions:""
    };
  }

  toggleContacts = id => {
    if(this.state.contacts.includes(id)) {
      let remainingContacts = this.state.contacts.filter(item => item !== id)
      this.setState({ contacts: remainingContacts })
    } else {
      this.setState({ contacts: [...this.state.contacts, id] })
    }
  };

  toggleContactsModal = () => {
    this.setState({ contacts_modal: !this.state.contacts_modal });
  };

  setVehicle = id => {
    this.setState({ vehicle: id });
    this.toggleVehicleModal();
  };

  toggleVehicleModal = () => {
    this.setState({ vehicle_modal: !this.state.vehicle_modal });
  };

  toggleGear = id => {
    if (this.state.gear.includes(id)) {
      let remainingItems = this.state.gear.filter(item => item !== id)
      this.setState({gear: remainingItems})
    } else {
      this.setState({gear: [...this.state.gear, id]})
    }
  };

  toggleGearModal = () => {
    this.setState({ gear_modal: !this.state.gear_modal });
  };

  handleSubmit = async () => {
    const { navigation, user, setCurrentTrip, setTrips } = this.props;
    const newTrip = {
      name: this.state.name,
      startingPoint: this.state.startingPoint,
      endingPoint: this.state.endingPoint,
      startDate: this.state.startDate,
      startTime: this.state.startTime,
      endDate: this.state.endDate,
      endTime: this.state.endTime,
      notificationDate: this.state.notificationDate,
      notificationTime: this.state.notificationTime,
      travelingCompanions: this.state.travelingCompanions,
      userId: user.id
    }
    const trip = await addTrip(newTrip);
    const tripId = trip.data.createTrip.trip.id;
    this.props.setCurrentTrip(tripId);
    const userInfoTrips = await getTrips();
    await setTrips(userInfoTrips.user.trips);
    this.saveTripGear(tripId);
    this.saveTripContact(tripId);
    this.saveTripVehicle(tripId);
    navigation.navigate("TripList");
    this.clearInputs();
  }

  saveTripGear = (tripId) => {
    let promises = this.state.gear.map(gearId => {
      let newGear = {
        tripId,
        gearId,
        comment: ""
      };
      addGearForTrip(newGear);
    })
    return Promise.all(promises);
  }

  saveTripContact = (tripId) => {
    let promises = this.state.contacts.map(emergencyContactId => {
      let newContact = {
        tripId,
        emergencyContactId
      };
      addContactsForTrip(newContact);
    })
    return Promise.all(promises);
  }

  saveTripVehicle = async (tripId) => {
    let newVehicle = {
      tripId,
      vehicleId: this.state.vehicle
    };
    await addVehiclesForTrip(newVehicle);
  }

  clearInputs = () => {
    this.setState({
      contact: 0,
      contact_modal: false,
      vehicle: 0,
      vehicle_modal: false,
      gear: [],
      gear_modal: false,
      name: "",
      startingPoint: "",
      endingPoint: "",
      startDate: "",
      startTime: "",
      endDate: "",
      endTime: "",
      notificationDate: "",
      notificationTime: "",
      travelingCompanions: ""
    });
  }

  render() {
    const disableBtn = this.props.currentTrip ? true : false;
    const disableBtnColor = this.props.currentTrip ? styles.disableColor : styles.updateBtn;
    const contactsList = this.props.contacts.map(contact => {
      return (
        <TouchableHighlight
          key={contact.id}
          style={styles.modalButton}
          onPress={() => this.toggleContacts(contact.id)}
        >
          <View
            style={styles.modalToggleContactsContainer}
          >
            <Text
              style={styles.modalToggleContactsBtn}>
              {!this.state.contacts.includes(contact.id) ? "ADD" : "REMOVE"}
            </Text>
            <Text
              style={styles.modalToggleContactsHeading}
            >{contact.name}</Text>
          </View>
        </TouchableHighlight>
      );
    });

    const vehiclesList = this.props.vehicles.map(vehicle => {
        return (
          <TouchableHighlight
            key={vehicle.id}
            style={styles.modalButton}
            onPress={() => this.setVehicle(vehicle.id)}
          >
            <Text>
              {vehicle.make} {vehicle.model}
            </Text>
          </TouchableHighlight>
        );
      });

    const gearList = this.props.gear.map(gearItem => {
      return (
        <TouchableHighlight
          key={gearItem.id}
          style={styles.modalButton}
          onPress={() => this.toggleGear(gearItem.id)}
        >
          <View
            style={styles.modalToggleGearContainer}
            >
            <Text 
              style={styles.modalToggleGearBtn}>
              {!this.state.gear.includes(gearItem.id) ? "ADD" : "REMOVE"}
            </Text>
            <Text
              style={styles.modalToggleGearHeading}
              >{gearItem.itemName}</Text>
          </View>
        </TouchableHighlight>
      );
    });

  return (
    <View style={styles.tripContainer}>
      <ScrollView style={styles.tripsList}>
        <TouchableOpacity
          style={styles.modalToggleButton}
          onPress={() => this.toggleContactsModal()}
        >
          <Text style={styles.modalToggleText}>Select Emergency Contacts</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={styles.modalToggleButton}
          onPress={() => this.toggleVehicleModal()}
        >
          <Text style={styles.modalToggleText}>Select Vehicle</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={styles.modalToggleButton}
          onPress={() => this.toggleGearModal()}
        >
          <Text style={styles.modalToggleText}>Select Gear Items</Text>
        </TouchableOpacity>
        <Text style={styles.label}>Trip Name: </Text>
        <TextInput
          placeholder="Mt Massive w/soccer club"
          style={styles.input}
          onChangeText={text => this.setState({ name: text })}
          value={this.state.name}
        />
        <Text style={styles.label}>Starting Point:</Text>
        <TextInput
          placeholder="Leadville Fish Hatchery"
          style={styles.input}
          onChangeText={text => this.setState({ startingPoint: text })}
          value={this.state.startingPoint}
        />
        <Text style={styles.label}>Ending Point:</Text>
        <TextInput
          placeholder="Leadville Fish Hatchery"
          style={styles.input}
          onChangeText={text => this.setState({ endingPoint: text })}
          value={this.state.endingPoint}
        />
        <Text style={styles.label}>Starting date:</Text>
        <TextInput
          placeholder="November 20, 2019"
          style={styles.input}
          onChangeText={text => this.setState({ startDate: text })}
          value={this.state.startDate}
        />
        <Text style={styles.label}>Start time:</Text>
        <TextInput
          placeholder="07:00"
          style={styles.input}
          onChangeText={text => this.setState({ startTime: text })}
          value={this.state.startTime}
        />
        <Text style={styles.label}>End date:</Text>
        <TextInput
          placeholder="November 20, 2019"
          style={styles.input}
          onChangeText={text => this.setState({ endDate: text })}
          value={this.state.endDate}
        />
        <Text style={styles.label}>End time:</Text>
        <TextInput
          placeholder="11:30"
          style={styles.input}
          onChangeText={text => this.setState({ endTime: text })}
          value={this.state.endTime}
        />
        <Text style={styles.label}>Notification date:</Text>
        <TextInput
          placeholder="November 20, 2019"
          style={styles.input}
          onChangeText={text => this.setState({ notificationDate: text })}
          value={this.state.notificationDate}
        />
        <Text style={styles.label}>Notification time:</Text>
        <TextInput
          placeholder="14:00"
          style={styles.input}
          onChangeText={text => this.setState({ notificationTime: text })}
          value={this.state.notificationTime}
        />
        <Text style={styles.label}>Number of Companions:</Text>
        <TextInput
          placeholder="3"
          style={styles.input}
          onChangeText={text => this.setState({ travelingCompanions: text })}
          value={this.state.travelingCompanions}
        />
        <Modal
          animationType={"slide"}
          visible={this.state.contacts_modal}
          transparent={true}
          onRequestClose={() => console.log("close requested")}
        >
          <View style={styles.pickerView}>
            <Text style={styles.modalHeading}>Add Contacts Items:</Text>
            <ScrollView>{contactsList}</ScrollView>
            <TouchableOpacity onPress={() => this.toggleContactsModal()}>
              <Text style={styles.updateBtn}>Submit Contacts List</Text>
            </TouchableOpacity>
          </View>
        </Modal>
        <Modal
          animationType={"slide"}
          visible={this.state.vehicle_modal}
          transparent={true}
          onRequestClose={() => console.log("close requested")}
        >
          <View style={styles.pickerView}>
            <Text style={styles.modalHeading}>Select Vehicle:</Text>
            {vehiclesList}
          </View>
        </Modal>
        <Modal
          animationType={"slide"}
          visible={this.state.gear_modal}
          transparent={true}
          onRequestClose={() => console.log("close requested")}
        >
          <View style={styles.pickerView}>
            <Text style={styles.modalHeading}>Add Gear Items:</Text>
            <ScrollView>{gearList}</ScrollView>
            <TouchableOpacity onPress={() => this.toggleGearModal()}>
              <Text style={styles.updateBtn}>Submit Gear List</Text>
            </TouchableOpacity>
          </View>
        </Modal>
        <TouchableOpacity
          style={disableBtnColor}
          onPress={this.handleSubmit}
          disabled={disableBtn}
        >
          <Text style={styles.updateBtnText}>Add Trip</Text>
        </TouchableOpacity>
      </ScrollView>
    </View>
  );
  }
}

export const mapStateToProps = ({ contacts, vehicles, gear, user, currentTrip }) => ({
  contacts,
  vehicles,
  gear,
  user,
  currentTrip
});

export const mapDispatchToProps = dispatch => bindActionCreators({ setTrips, setCurrentTrip }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Trip);
