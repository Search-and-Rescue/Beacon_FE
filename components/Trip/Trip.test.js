import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import { Trip } from './Trip';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

jest.mock("react-native-gesture-handler", () => "TouchableOpacity");

describe('Trip', () => {
  let wrapper;
  const contactsMock = [{
    name: "Beans",
    phone: "303-1234",
    email: "beans@gmail.com"
  }];
  const vehiclesMock = [{
    id: 1,
    make: 'Ford',
    model: 'F150',
    year: 2010,
    color: 'silver',
    licensePlate: 'CYE 909'
  }];
  const gearMock = [{
    id: 89,
    itemName: 'Rustic Iron Bottle',
    description: 'Chuck Norris can binary search unsorted data.'
  }];

  beforeEach(() => {
    mockProps = {
      navigation: {
        state: {
          params: {
            gearArray: {
              id: 1,
              gear_name: "tent",
              description: "REI red 2 person"
            }
          }
        }
      }
    };

    wrapper = shallow(<Trip 
      contacts={contactsMock}
      vehicles={vehiclesMock}
      gear={gearMock}
      {...mockProps} />)
  });

  it('should match the snapshot with all of the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });
});