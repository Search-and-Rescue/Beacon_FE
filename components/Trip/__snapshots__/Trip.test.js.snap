// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trip should match the snapshot with all of the data passed through 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 2,
        "flex": 1,
        "height": 300,
        "width": "90%",
      }
    }
  >
    <Text>
      State is 
      0
    </Text>
    <Button
      onPress={[Function]}
      title="Add Emergency Contact"
    />
    <Text>
      State is 
      0
    </Text>
    <Button
      onPress={[Function]}
      title="Add Vehicle"
    />
    <Text>
      State length 
      0
    </Text>
    <Button
      onPress={[Function]}
      title="Add Gear"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="Starting point"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="Ending point"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="Start date"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="Start time"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="End date"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="End time"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      placeholder="Notification time"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "black",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <Component
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "bottom": 30,
            "height": 400,
            "left": 10,
            "margin": 5,
            "padding": 20,
            "position": "absolute",
            "right": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 5,
            }
          }
        >
          Select Emergency Contact:
        </Text>
        <TouchableHighlight
          activeOpacity={0.85}
          delayPressOut={100}
          onPress={[Function]}
          style={
            Object {
              "marginVertical": 3,
              "paddingVertical": 10,
            }
          }
          underlayColor="black"
        >
          <Text>
            Beans
          </Text>
        </TouchableHighlight>
      </View>
    </Component>
    <Component
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "bottom": 30,
            "height": 400,
            "left": 10,
            "margin": 5,
            "padding": 20,
            "position": "absolute",
            "right": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 5,
            }
          }
        >
          Select Vehicle:
        </Text>
        <TouchableHighlight
          activeOpacity={0.85}
          delayPressOut={100}
          key="1"
          onPress={[Function]}
          style={
            Object {
              "marginVertical": 3,
              "paddingVertical": 10,
            }
          }
          underlayColor="black"
        >
          <Text>
            Ford
             
            F150
          </Text>
        </TouchableHighlight>
      </View>
    </Component>
    <Component
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "bottom": 30,
            "height": 400,
            "left": 10,
            "margin": 5,
            "padding": 20,
            "position": "absolute",
            "right": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 5,
            }
          }
        >
          Add Gear Items:
        </Text>
        <Component>
          <TouchableHighlight
            activeOpacity={0.85}
            delayPressOut={100}
            key="89"
            onPress={[Function]}
            style={
              Object {
                "marginVertical": 3,
                "paddingVertical": 10,
              }
            }
            underlayColor="black"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Object {
                    "borderColor": "black",
                    "borderWidth": 1,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingVertical": 3,
                    "textAlign": "center",
                    "width": 75,
                  }
                }
              >
                ADD
              </Text>
              <Text
                style={
                  Object {
                    "paddingVertical": 10,
                  }
                }
              >
                Rustic Iron Bottle
              </Text>
            </View>
          </TouchableHighlight>
        </Component>
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Text
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 1,
                "marginVertical": 3,
                "padding": 10,
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            Submit Gear List
          </Text>
        </TouchableOpacity>
      </View>
    </Component>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "marginVertical": 3,
          "padding": 10,
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Add Trip
      </Text>
    </TouchableOpacity>
  </Component>
</View>
`;
